jgeorge37/acmwosu-db
# Jcg/ada 19 password reset token

(This is just for the backend, I have another story in the backlog to tie together all of the sign-in and password reset pieces)

This is to generate a password reset token - I.e. once everything comes together, the user can submit their email address to reset their password. When they do this, my API endpoint checks that the email address is associated with an account, and if so updates the reset_token column with a randomized string and generates an expiration timestamp (2 days in the future). This token will be used in the link that the user gets emailed to them. The token is a unique identifier for the user for the 2 days.

This change includes a database migration so even if you are not reviewing, once this gets merged you should npm run migrate on your local db.

Part 1 - To test the token generation:
0. Have a local DB (you know the drill) and something like Postman that is capable of doing post requests.
1. `npm run migrate` to update your db with the new columns
1. Add an account to your account table if your account table is empty (I would use psql for this personally)
2. `npm run dev`
3. Send a post request to http://localhost:3000/api/account/generate-token with the body containing "email": "emailThatYouHaveAnAccountFor@whatever.com" (but obvi with the actual email you made an account with in the account table)
4. Check that the token was created and that the expiration date is 2 days in the future (SELECT * FROM account;)
5. Try hitting the endpoint again with an email that has no associated account; you should get an relevant error message in the request response.

Part 2 - To test the password reset + token expiration:
0. `npm run dev`
1. Copy the generated token (prolly from psql) for the account you just did the above for.
2. Send a post request to http://localhost:3000/api/account/reset-pw with the body "token": THE TOKEN YOU COPIED and "password" THE NEW PASSWORD - dont keyboard smash bc in the db it will be encrypted so you have to remember it
3. Try the sign in page with the new password, you should not get a message about a bad password
4. To test expiration, set the time column for the account you have been messing with to the current time: UPDATE account SET token_expire_time = current_timestamp WHERE email = THE EMAIL OF THE ACCOOUNT
5. Try resetting the password again with Postman, using the same token. The response should contain a relevant error message.

and try to poke more holes if you want
